modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/metadata/MetadataKeyIdObjectResolver.java

```diff
@@ -127,6 +132,58 @@ public Object resolve() throws MetadataResolvingException {
     return visitor.getResultId();
   }
 
+  private MetadataKeyBuilder getKeyFromField(Object resolvedKey, DeclaringMemberModelProperty declaringMemberModelProperty)
+      throws Exception {
+    return newKey(valueOf(getFieldValue(resolvedKey, declaringMemberModelProperty.getDeclaringField().getName())));
+  }
+
+  /**
+   * Given a {@link Object} representing the resolved value for a {@link MetadataKey}, generates the {@link MetadataKey} object.
+   * 
+   * @param resolvedKey
+   * @return {@link MetadataKey} reconstructed from the resolved object key
+   * @throws MetadataResolvingException
+   */
+  MetadataKey reconstructKeyFromType(Object resolvedKey) throws MetadataResolvingException {
+    if (isKeyLess()) {
+      return new NullMetadataKey();
+    }
+
+    if (keyParts.size() == 1) {
+      return newKey(valueOf(resolvedKey)).build();
+    }
+
+    Reference<MetadataKeyBuilder> rootBuilder = new Reference<>();
```

Comment:
3 reference objects just to be able to use a stream is not worth it. Do old school iteration instead and have a cleaner code